### PR TITLE
add broadcasting fusion

### PR DESCRIPTION
**Pull request by Moelf**
On 1.7-rc1:
```julia
julia> testf(evt) = evt.nMuon==4
testf (generic function with 1 method)

julia> testf2(evt) = evt.nMuon==4
testf2 (generic function with 1 method)

julia> function naive(t)
           a1 = testf.(t)
           a2 = testf2.(t)
           findall(a1 .&& a2)
       end
naive (generic function with 1 method)

julia> @time naive(tf);
  1.620120 seconds (28.63 k allocations: 1.679 GiB, 7.97% gc time)

julia> @time findall(@. testf(tf) && testf2(tf));
  0.930162 seconds (387.13 k allocations: 893.470 MiB, 0.40% gc time, 14.64% compilation time)

julia> naive(tf) == findall(@. testf(tf) && testf2(tf))
true
```

### CODE REVIEW

File: src/iteration.jl [hunk 1]
```diff
@@ -195,15 +195,19 @@ end
 Base.propertynames(lt::LazyTree) = propertynames(innertable(lt))
 Base.getproperty(lt::LazyTree, s::Symbol) = getproperty(innertable(lt), s)
 
-# a specific branch
+Base.broadcastable(lt::LazyTree) = lt
+Base.IndexStyle(::Type{<:LazyTree}) = IndexLinear()
+Base.findall(testf::Function, lt::LazyTree) = [testf(evt) for evt in lt]
```
Comment (aminnj) [inline src/iteration.jl:200]: "testf" -> "f"?

Comment (Moelf) [reply src/iteration.jl:200]:
```suggestion
Base.findall(f::Function, lt::LazyTree) = [f(evt) for evt in lt]
```

Comment (Moelf) [reply src/iteration.jl:200]: I can't apply my own suggestion :(

File: src/iteration.jl [hunk 2]
```diff
@@ -195,15 +195,19 @@ end
 Base.propertynames(lt::LazyTree) = propertynames(innertable(lt))
 Base.getproperty(lt::LazyTree, s::Symbol) = getproperty(innertable(lt), s)
 
-# a specific branch
+Base.broadcastable(lt::LazyTree) = lt
+Base.IndexStyle(::Type{<:LazyTree}) = IndexLinear()
+Base.findall(f::Function, lt::LazyTree) = [f(evt) for evt in lt]
```
Comment (aminnj) [inline src/iteration.jl:200]: Can this be `f.(lt)`? Then it returns a more compact BitVector instead of Vector{Bool}

Comment (aminnj) [reply src/iteration.jl:200]: Ah wait, is this the same behavior as `findall([true, false, true, false])` which gives `[1,3]`?

Comment (Moelf) [reply src/iteration.jl:200]: wait, you're right, this is not normal `findall`

Comment (aminnj) [reply src/iteration.jl:200]: Base has
```julia
findall(testf::Function, A) = collect(first(p) for p in pairs(A) if testf(last(p)))

# Broadcasting is much faster for small testf, and computing
# integer indices from logical index using findall has a negligible cost
findall(testf::Function, A::AbstractArray) = findall(testf.(A))
```

Comment (aminnj) [reply src/iteration.jl:200]: If normal broadcasting works now, then the Base implementation should just work

**Comment by aminnj** [comment]
Is this expected (using this PR)?
```julia
julia> using UnROOT;const tf = LazyTree(ROOTFile("lz4_Run2012BC_DoubleMuParked_Muons.root"),"Events");
[ Info: Precompiling UnROOT [3cd96dde-e98d-4713-81e9-a4a1b0235ce9]

julia> myfilt(evt) = evt.nMuon == 4

julia> @time myfilt.(tf)
 84.448891 seconds (923.14 M allocations: 95.784 GiB, 15.69% gc time)
61540413-element BitVector:
 0
 0

julia> @time map(myfilt,tf);
  0.375699 seconds (2.50 k allocations: 928.657 MiB, 4.74% gc time)

julia> @time [myfilt(evt) for evt in tf];
  0.393586 seconds (50.61 k allocations: 931.106 MiB, 3.53% gc time, 8.92% compilation time)
```

**Comment by aminnj** [comment]
Also we should test for the table pretty printing explicitly (not just that it doesn't crash) since with the PR it got broken but passes CI
```julia
julia> tf
61540413-element LazyTree{TypedTables.Table{NamedTuple{(:Muon_phi, :nMuon, :Muon_pt, :Muon_eta, :Muon_charge, :Muon_mass), Tuple{Vector{Float32}, UInt32, Vector{Float32}, Vector{Float32}, Vector{Int32}, Vector{Float32}}}, 1, NamedTuple{(:Muon_phi, :nMuon, :Muon_pt, :Muon_eta, :Muon_charge, :Muon_mass), Tuple{LazyBranch{Vector{Float32}, UnROOT.Nooffsetjagg, ArraysOfArrays.VectorOfVectors{Float32, Vector{Float32}, Vector{Int32}, Vector{Tuple{}}}}, LazyBranch{UInt32, UnROOT.Nojagg, Vector{UInt32}}, LazyBranch{Vector{Float32}, UnROOT.Nooffsetjagg, ArraysOfArrays.VectorOfVectors{Float32, Vector{Float32}, Vector{Int32}, Vector{Tuple{}}}}, LazyBranch{Vector{Float32}, UnROOT.Nooffsetjagg, ArraysOfArrays.VectorOfVectors{Float32, Vector{Float32}, Vector{Int32}, Vector{Tuple{}}}}, LazyBranch{Vector{Int32}, UnROOT.Nooffsetjagg, ArraysOfArrays.VectorOfVectors{Int32, Vector{Int32}, Vector{Int32}, Vector{Tuple{}}}}, LazyBranch{Vector{Float32}, UnROOT.Nooffsetjagg, ArraysOfArrays.VectorOfVectors{Float32, Vector{Float32}, Vector{Int32}, Vector{Tuple{}}}}}}}}:
 (Muon_phi = Float32[-0.034272723, 2.5426154], nMuon = 0x00000002, Muon_pt = Float32[10.763697, 15.736523], Muon_eta = Float32[1.0668273, -0.5637865], Muon_charge = Int32[-1, -1], Muon_mass = Float32[0.10565837, 0.10565837])
 (Muon_phi = Float32[-0.2747921, 2.5397813], nMuon = 0x00000002, Muon_pt = Float32[10.53849, 16.327097], Muon_eta = Float32[-0.42778006, 0.34922507], Muon_charge = Int32[1, -1], Muon_mass = Float32[0.10565837, 0.10565837])
# ...
```

**Comment by Moelf** [comment]
>julia> @time myfilt.(tf) # materializing everything
 84.448891 seconds (923.14 M allocations: 95.784 GiB, 15.69% gc time)

ah no, this is not intended, broadcasting should use `LazyEvent`.

**Comment by Moelf** [comment]
the easy fix is to make
```diff
-Base.getindex(lt::LazyTree, row::Int) = innertable(lt)[row]
+Base.getindex(lt::LazyTree, row::Int) = LazyEvent(innertable(lt), row)
```

**Comment by Moelf** [comment]
```julia
julia> tf[1]
UnROOT.LazyEvent at index 1 with 2 columns:
(nMuon = 0x00000002, Muon_pt = Float32[10.763697, 15.736523])

julia> @time myfilt.(tf);
  0.760156 seconds (14.28 k allocations: 836.077 MiB, 4.07% gc time)

julia> tf
 Row │ nMuon   Muon_pt                              
     │ UInt32  Vector{Float32}                      
─────┼──────────────────────────────────────────────
 1   │ 2       [10.8, 15.7]
 2   │ 2       [10.5, 16.3]
 3   │ 1       [3.28]
 4   │ 4       [11.4, 17.6, 9.62, 3.5]
 5   │ 4       [3.28, 3.64, 32.9, 23.7]
 6   │ 3       [3.57, 4.57, 4.37]
 7   │ 2       [57.6, 53.0]
 8   │ 2       [11.3, 23.9]
```

File: Project.toml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 name = "UnROOT"
 uuid = "3cd96dde-e98d-4713-81e9-a4a1b0235ce9"
 authors = ["Tamas Gal", "Jerry Ling", "Johannes Schumann", "Nick Amin"]
-version = "0.6.3"
```
Comment (Moelf) [inline Project.toml:4]: actually this should be a breaking change

File: Project.toml [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
+version = "0.6.4"
```
Comment (Moelf) [inline Project.toml:4]:
```suggestion
version = "0.7.0"
```

**Review by tamasgal — APPROVED** [review]

**Comment by tamasgal** [comment]
I like it `:)` I guess it's time to collect some benchmark lines for the regression tests (https://github.com/tamasgal/UnROOT.jl/issues/127)

**Comment by aminnj** [comment]
I like the lazy-until-show trick 😉

**Pull request merged**